src/core/algorithms/fd/fd_verifier/dynamic_fd_verifier.cpp

```diff
@@ -0,0 +1,247 @@
+            }
+            if (rows_to_delete.contains(id)) {
+                throw config::ConfigurationError("Delete statements have duplicates");
+            }
+            rows_to_delete.emplace(id);
+        }
+    };
+
+    auto check_updates = [this](config::InputTable update_batch) {
+        if (!update_batch->HasNextRow()) {
+            return;
+        }
+        if (update_batch->GetNumberOfColumns() != input_table_->GetNumberOfColumns() + 1) {
+            throw config::ConfigurationError(
+                    "Update statements has differ column number from input table");
+        }
+        for (size_t i = 0; i < input_table_->GetNumberOfColumns(); ++i) {
+            if (update_batch->GetColumnName(i + 1) != input_table_->GetColumnName(i)) {
+                throw config::ConfigurationError(
+                        "Update statements has differ columns from input table");
+            }
+        }
+        std::unordered_set<size_t> rows_to_update;
+        while (update_batch->HasNextRow()) {
+            auto row = update_batch->GetNextRow();
+            size_t id = std::stoull(row[0]);
+            if (!stats_calculator_->ValidateDeletingRow(id)) {
+                throw config::ConfigurationError("Attempt to update a non-existing row");
+            }
+            if (rows_to_update.contains(id)) {
+                throw config::ConfigurationError("Update statements have duplicates");
+            }
+            rows_to_update.emplace(id);
+        }
+        update_batch->Reset();
+    };
+
+    RegisterOption(config::kTableOpt(&input_table_));
+    RegisterOption(config::kInsertStatementsOpt(&insert_batch_).SetValueCheck(check_inserts));
+    RegisterOption(config::kDeleteStatementsOpt(&delete_batch_).SetValueCheck(check_deletes));
+    RegisterOption(config::kUpdateStatementsOpt(&update_batch_).SetValueCheck(check_updates));
+    RegisterOption(config::kUpdateInPlaceOpt(&update_in_place_));
+    RegisterOption(config::kLhsIndicesOpt(&lhs_indices_, get_schema_cols));
+    RegisterOption(config::kRhsIndicesOpt(&rhs_indices_, get_schema_cols));
+}
+
+void DynamicFDVerifier::MakeExecuteOptsAvailable() {
+    using namespace config::names;
+    MakeOptionsAvailable(kCrudOptions);
+}
+
+void DynamicFDVerifier::LoadDataInternal() {
+    if (!input_table_->HasNextRow()) {
+        throw std::runtime_error("Got an empty dataset: FD verifying is meaningless.");
+    }
+    CreateFD();
+    input_table_->Reset();
+    stats_calculator_ =
+            std::make_unique<DynamicStatsCalculator>(input_table_, lhs_indices_, rhs_indices_);
+    VerifyFD();
+    SortHighlightsByProportionDescending();
+}
+
+unsigned long long DynamicFDVerifier::ExecuteInternal() {
+    auto start_time = std::chrono::system_clock::now();
+    std::vector<std::pair<std::optional<size_t>, std::vector<int>>> lhs_inserts{}, rhs_inserts{};
+    std::vector<size_t> lhs_deletes{delete_batch_}, rhs_deletes{delete_batch_};
```

Comment:
`lhs_deletes` and `rhs_deletes` always have the same contents, don't separate them. In fact, passing `delete_batch_` straight to the methods might work even better.

Reply:
> `delete_batch_` should be an `unordered_set` too.
I think it's shouldn't and we don't need `unordered_set` at all, because we don't do any lookup through `delete_batch_`, we only iterate over it